README.md
# ⚠️ NO LONGER MAINTAINED ⚠️
If you can fix the bot and get it running again, I will gladly welcome a pull request, but I just do not have the cognitive bandwidth between finals, the protests, and my upcoming internship to maintain the bot now that it no longer seems to work. I apologize.

# COVID-19 Tracker v10.7d
Bot for posting information on spread of Coronavirus Disease 2019 (COVID-19) to Twitter and Discord. The official instance of this bot can be found at [@COVID19Tracker](https://twitter.com/COVID19Tracker).

# To create your own instance:

1. Sign up for Twitter API access [here](https://developer.twitter.com/) and make note of your consumer key, consumer secret, access token, and access token secret. _TWITTER WILL ONLY SHOW YOU YOUR ACCESS TOKEN AND ACCESS TOKEN SECRET **ONCE,** SO DON'T LOSE THEM._ **DO NOT SHARE THESE WITH ANYONE OR THEY CAN TAKE OVER YOUR TWITTER ACCOUNT.**

2. Make a `credentials.py` file in the following format:

```
import tweepy

consumer_key = ''
consumer_secret = ''
access_token = ''
access_token_secret = ''
```

3. Run `Setup-nCoV.sh`. This will create your Python 3 virtual environment, install all required packages, create the launcher scripts to run the program without having to activate the virtual environment, create the cron job for the bot to run automatically, and test the script for you in notweet mode.

4. Create your `config.txt` file, using config_example.txt as a base. The config file requires 9 entries to function correctly, in the following format:

```
[FLAG EMOJI]
[DICTIONARY KEY FOR DESIRED COUNTRY]
[OUTPUT NAME FOR COUNTRY]
[MINIMUM NUMBER OF DAYS TO CLIP OFF OF LEFT SIDE OF GRAPH]
[FILTER LIST] (comma-separated keylist if a 'country' is made up of multiple countries - NOT FOR 'World')
0
0
0
```

I advise you put the `World` key in there somewhere, for obvious reasons.

5. This bot can also post to Discord. In your Discord server settings, generate webhooks for whatever channels you want the bot to post in, and place each webhook URL on its own line in a `webhooks.txt` file.

Please note the bot will automatically overwrite your Twitter bio with a 'last updated' statistic. If you have a bio you want to use in addition to this, please create 'bio_top.txt' and 'bio_bottom.txt' files containing your desired information. The last updated count will appear between the text contained in these two files, both of which are optional.

You're done! The bot should now be working.

# About
Cases, dead, and recovered counts are from the [Johns Hopkins University CSSE tracker](https://gisanddata.maps.arcgis.com/apps/opsdashboard/index.html#/bda7594740fd40299423467b48e9ecf6).'

This link leads to dashboards with visual maps as well as per-state statistics, in English, and sources information from the [WHO](https://www.who.int/emergencies/diseases/novel-coronavirus-2019/situation-reports), [CDC](https://www.cdc.gov/coronavirus/2019-ncov/index.html), [ECDC](https://www.ecdc.europa.eu/en/geographical-distribution-2019-ncov-cases), [NHC](http://www.nhc.gov.cn/xcs/yqtb/list_gzbd.shtml), [DXY](https://ncov.dxy.cn/ncovh5/view/pneumonia?scene=2&clicktime=1579582238&enterid=1579582238&from=singlemessage&isappinstalled=0), [1Point3Acres](https://coronavirus.1point3acres.com/), [worldometers.info](https://www.worldometers.info/coronavirus/), [BNO News](https://bnonews.com/index.php/2020/04/the-latest-coronavirus-cases/), and [The COVID Tracking Project](https://covidtracking.com/data). It appears to be updated by hand, and thus are not 'up-to-the-minute'.

Active cases count is derived by the following formula: (cases - dead - recovered = active)

Updates are time-and-datestamped based on time of data being fetched and posted. Time-and-datestamp does NOT refer to when the information was last updated. For that, see the links above.

Tweets are in the following format:

```
🕓MM/DD ##H #COVID19 #Coronavirus

🏳️😷X,XXX,XXX 🏥X,XXX,XXX 💀XXX,XXX 👍XXX,XXX
🏳️😷XXX,XXX 🏥XXX,XXX 💀XX,XXX 👍XXX,XXX
🏳️😷XXX,XXX 🏥XXX,XXX 💀XX,XXX 👍XXX,XXX
🏳️😷XXX,XXX 🏥XXX,XXX 💀XX,XXX 👍XXX,XXX
🏳️😷XXX,XXX 🏥XXX,XXX 💀XX,XXX 👍XXX,XXX

😷Total 🏥Active 💀Dead 👍Recovered

[IMAGE OF TIME-SERIES GRAPHS]
```

This bot also optionally can output to Discord in the form of embeds in webhook messages. Discord embeds are in the following format:

```
⚠️   Coronavirus Update   ⚠️

🏳️ [COUNTRY NAME]:	🏳️ [COUNTRY NAME]:	🏳️ [COUNTRY NAME]:			
😷 X,XXX,XXX (+XX,XXX)	😷 X,XXX,XXX (+XX,XXX)	😷 X,XXX,XXX (+XX,XXX)	
🏥 X,XXX,XXX (+XX,XXX)	🏥 X,XXX,XXX (+XX,XXX)	🏥 X,XXX,XXX (+XX,XXX)	
💀 XXX,XXX (+XX,XXX)	💀 XXX,XXX (+XX,XXX)	💀 XXX,XXX (+XX,XXX)		
👍 XXX,XXX (+XX,XXX)	👍 XXX,XXX (+XX,XXX)	👍 XXX,XXX (+XX,XXX)		

🏳️ [COUNTRY NAME]:	🏳️ [COUNTRY NAME]:	🏳️ [COUNTRY NAME]:			
😷 X,XXX,XXX (+XX,XXX)	😷 X,XXX,XXX (+XX,XXX)	😷 X,XXX,XXX (+XX,XXX)	
🏥 X,XXX,XXX (+XX,XXX)	🏥 X,XXX,XXX (+XX,XXX)	🏥 X,XXX,XXX (+XX,XXX)	
💀 XXX,XXX (+XX,XXX)	💀 XXX,XXX (+XX,XXX)	💀 XXX,XXX (+XX,XXX)		
👍 XXX,XXX (+XX,XXX)	👍 XXX,XXX (+XX,XXX)	👍 XXX,XXX (+XX,XXX)		

🏳️ [COUNTRY NAME]:	🏳️ [COUNTRY NAME]:	🏳️ [COUNTRY NAME]:			
😷 X,XXX,XXX (+XX,XXX)	😷 X,XXX,XXX (+XX,XXX)	😷 X,XXX,XXX (+XX,XXX)	
🏥 X,XXX,XXX (+XX,XXX)	🏥 X,XXX,XXX (+XX,XXX)	🏥 X,XXX,XXX (+XX,XXX)	
💀 XXX,XXX (+XX,XXX)	💀 XXX,XXX (+XX,XXX)	💀 XXX,XXX (+XX,XXX)		
👍 XXX,XXX (+XX,XXX)	👍 XXX,XXX (+XX,XXX)	👍 XXX,XXX (+XX,XXX)		

[IMAGE OF TIME-SERIES GRAPHS]

😷Total 🏥Active 💀Dead 👍Recovered

🕓 Last Updated: YYYY-MM-DD HH:MM UTC
```

In both cases, updates will contain graphs showing up to the last approximately 11 weeks of confirmed cases, active cases, deaths, and recoveries in your chosen countries, with a monimum bound of days set by your config file.

This bot will, if functioning normally, check for new data once every two hours, and only post when there is new data. If any problems arise, please report them to @Ellie the Tinkerer#7777 on Discord.

# Credits

Credit to [@Yuu6883](https://github.com/Yuu6883) for helping write the web scraping code.

Credit to users on the [Weaponsandstuff93](https://www.youtube.com/channel/UCAbwEStxHetWMGvaq9FIF_w) Discord server for helping devise the post format in early versions of the bot.

Credit to Ensheng "Frank" Dong ([@Energeticodefish](https://github.com/enshengdong) on Github) from the Center for Systems Science and Engineering at Johns Hopkins University for replying to my numerous emails concerning the JHU CSSE tracker.

Credit to the CDC for the official instance's Twitter profile picture.

This bot is powered by Python 3 and the [Requests](https://requests.readthedocs.io/en/master/), [Tweepy](https://www.tweepy.org/), [pandas](https://pandas.pydata.org/), [discord-webhook](https://github.com/lovvskillz/python-discord-webhook), and [matplotlib](https://matplotlib.org/) libraries.

Data tracked by this bot is Copyright 2020 Johns Hopkins University.

# Terms of Use

The Johns Hopkins University data tracked by this bot is provided to the public strictly for educational and academic research purposes. The data is built from publicly available data from multiple sources, that do not always agree. Both I and Johns Hopkins University hereby disclaim any and all representations and warranties with respect to the bot and related data, including accuracy, fitness for use, and merchantability. Reliance on the bot or data for medical guidance or use of the data in commerce is strictly prohibited by both me and Johns Hopkins University.


CHANGELOG.md
# Changelog

## 10.7x:
- 10.7d
    - Removed mention of unfinished/removed 'verbose output' feature from README.md
    - Bumped version (forgot to do this for v10.7c release)
    - No other changes, project is still no longer being maintained
- 10.7c
    - Maintenance release to bump urllib3 and NumPy versions for security reasons (untested but should be fine)
    - Also updated contact information in the readme
    - No other changes, project is still no longer being maintained
- 10.7b
    - Reverted reduced import statements; turns out this doesn't actually help anything. Lesson learned.
- 10.7a
    - Fixed minor mistake in README.md, otherwise identical
- 10.7
    - Reduced import statements to bare minimum needed for functionality
        - I know this isn't best practices due to the risk of namespace collisions and code readability, but it shaves like 7 seconds off runtime and nothing broke

## 10.6x:
- 10.6
    - Changes to Setup-nCoV.sh to prevent cronjob mail spam

## 10.5x:
- 10.5a
    - Slight format fix to Setup-nCoV.sh, otherwise identical
- 10.5: CUMULATIVE UPDATE rolling up several versions of unreleased updates
    - 10.5:
        - Code is now fully commented
        - Fixed bug in tweet resending code not leading to proper exit condition
        - Fixed MAJOR bug causing inaccurate active case numbers in United States
    - 10.4:
        - Removed config file entries and unnecessary code fragments related to NumPy / PIL graph concatenation
    - 10.3:
        - No longer reliant on NumPy / PIL for graphs, code originally by JimChr - R4GN4R removed
    - 10.2:
        - Graph image now inside embed
    - 10.1-b:
        - Fixed resend-attempt code, fixed many bugs related to backend rework
    - 10.0-b:
        - MASSIVE backend rework to allow for changing output countries based on a config file instead of being hardcoded (country order changes no longer tracked in changelog)
    - 9.3-b
        - Fixed and reworked run logging (was not logging to file, was not logging manual bot execution)
        - Changed Discord output to [World, EU+UK, US, Spain, Italy, UK, Canada, South Korea, China], with graph output on both Twitter and Discord being a 3x3 composite of relevant graphs
    - 9.2-b
        - Added minor ticks to graph Y axis
    - 9.1-b
        - Number of days in graph now automatically adjusts each day from a per-graph minimum to a global maximum
        - Replaced Canada with Italy
    - 9.0-b1:
        - Fixed globe emoji displaying Europe/Africa instead of the Americas during American daytime hours
    - 9.0-b:
        - Reduced code repetition by breaking out Discord, Twitter, and graph generation into functions
        - Partially implemented new, rewritten Discord webhook (now uses inline fields in embed, 10.x non-beta release will put the image inside the embed as well)
    - 8.2-b
        - Discord webhook now outputs a composite graph of figures worldwide, in the US, in Canada, in the United Kingdom, in the EU, and in China (credit to JimChr - R4GN4R on StackOverflow for the code for this)
    - 8.1-b:
        - Added graph colors, reworked graph ticks/labels, added grid to graphs, generally cleaned up graphing code
        - Removed year from graph labels, increased graph from 'last 9 days' to 'last 12 days'
    - 8.0-b
        - Added graphs for worldwide, US, UK, and EU counts, removed code related to location-based updates, as feature is no longer on roadmap
    - 7.0-b1:
        - Replaced China with UK on Twitter output (both UK and China are retained on Discord output)
        - Removed deltas on Twitter output, changed tweet format to be even more compact to allow for a footer explaining how to read the new compact tweets
    - 7.0-b:
       - Rewrote how the bot fetches and handles JHU data, further compacted tweet, replaced 'Without China' count with US, Canada, EU, and China counts
    - 6.0-b2:
        - Compacted tweet format due to tweet length causing bot failure (code not formally 'released' as a specific build, pushed to GitHub as part of  release)

## 6.0-bx:
- 6.0-b1: Fixed Setup-nCoV.sh to use new footer format, removed Setup-nCoV.sh user input (no longer needed)
- 6.0-b: Added Discord webhook functionality, FINALLY fixed output log archiving

## 5.0-bx:
- 5.0-b: Added 'outside China' count, total rewrite of tweet format (code pushed to GitHub but not formally 'released' as a specific build)

## 4.2-beta-x:
- 4.2-b1: Major changes to Twitter-related logic to prevent bugs caused by Twitter being over-capacity.
- 4.2-beta: Added 'last updated' output to Twitter bio

## 4.1-beta-x:
- 4.1-beta: Removed Tencent, added 'active' count, fixed various bugs with data fetching (code pushed to GitHub but not formally 'released' as a specific build)

## 4.0-beta-x:
- 4.0-beta-1: Fixed setup script, stubbed out / removed Google Sheets API (verbose output currently hardcoded to produce no output pending csv update)
- 4.0-beta: Complete overhaul of front-facing UI, added full parsing engine and supporting code for location-based updates using Google Sheets (which was deprecated right as this release was being finalized, hence the beta and new version number, I need to rewrite it for GitHub-hosted .csv files)

## 3.2-beta-x
- 3.2-beta-1: CRITICAL bugfix in set-up script (cronjob called wrong file)
- 3.2-beta: Rewrote the setup procedure; aside from setting up API access and creating your `credentials.py` and `credentials.json` files, you should now only need to run the setup script for the bot to work. I also fixed Setup-nCoV.py creating `hashtags.txt` instead of `footers.txt`.

## 3.0-beta-x
- 3.0-beta-3: Replaced `hashtags.txt` with `footer.txt`, which defines both the 'Please retweet' message as well as the hashtags. This file is now created by the installer script. Fixed a MAJOR bug in 3.0-beta-2 where the bot would post `ABOT` instead of simply not posting. Fixed a behavior where the bot would only not produce a tweet if tweeting was disabled.

nCoV_twitter.py
#!/usr/bin/python3

# Coronavirus Disease Tracker v10.7d
# By Math Morissette (@TheYadda on Github)
# Last updated: 2020-04-15
#
# A Twitter/Discord bot for posting information on the spread of the COVID-19
# outbreak
#
# Uses Requests, Tweepy, pandas, discord-webhook, and matplotlib libraries
#
# File: nCoV_twitter.py
# Purpose: Methods for nCoV.py that use Twitter and Discord APIs

# Import our Twitter credentials and Tweepy library from credentials.py
from credentials import *

# For generating timestamps and delay when resending
import time
import random

# Import Discord webhook support
from discord_webhook import DiscordWebhook, DiscordEmbed

# Get Twitter API access
def get_twitter_api(key, secret, token, token_secret):
    # Authenticate to Twitter
    auth = tweepy.OAuthHandler(key, secret)
    auth.set_access_token(token, token_secret)

    # Create API object
    api = tweepy.API(auth, wait_on_rate_limit=True,
        wait_on_rate_limit_notify=True)
    
    return api

# Send nCoV tweets
def output(
    send_flag, api, tweet, stats_discord, date_discord, tweet_data, datecode, \
    hour, update_arr):

    # Logger output: tweet
    print(f"Tweet:\n{tweet}\n")

    # Send tweet
    if send_flag == True:
        # If an exception is thrown, try a total of three times; if it fails
        # all three times, raise excewption
        for attempt_no in range(0,4):
            try:
                # Add graphs to tweet
                filename = 'graph.png'
                media_ids = []
                res = api.media_upload(filename)
                media_ids.append(res.media_id)

                # Actually send tweet
                tweet_obj = api.update_status(status=tweet, \
                            media_ids=media_ids, tweet_mode='extended')

                # Write updated config file
                with open ("config.txt", "w") as tweet_file:
                    tweet_file.write("\n".join(map(str, update_arr)))
                    tweet_file.close()
                break
            except:
                if attempt_no < 3:
                    print("Retrying")
                    time.sleep(random.randint(15,30))
                else:
                    raise error
        
        # Update time last checked per Twitter profile
        lastcheckedupdate(tweet[0], send_flag, api, datecode)

    # Send Discord embeds, if webhooks file exists
    # Otherwise, skip Discord output
    with open ("webhooks.txt", "r") as webhook_file:
        webhook_urls = webhook_file.readlines()
        webhook_file.close()

        # Emoji translation key at bottom of embed
        emojikey = "😷Total 🏥Active 💀Dead 👍Recovered"

        # Create embed and populate it with case data
        embed = DiscordEmbed(color=16737792)
        for key in stats_discord:
            embed.add_embed_field(
                name=key, value=stats_discord[key], inline=True)
        embed.set_footer(text=f"{emojikey}\n{date_discord}")

        # If tweet was sent and embeds are being sent, attach image from it
        if send_flag == True:
            image_url = tweet_obj.entities['media'][0]['media_url']
            try:
                embed.set_image(url=f"{image_url}?format=png&name=4096x4096")
            except:
                pass

        # Attach tweet to embed name
        #
        # If no tweet was sent but script somehow proceeded (shouldn't be
        # possible), attach twitter account to embed name instead
        try:
            embed.set_author(
                name='⚠️ Coronavirus Update ⚠️', url=(f"https://twitter.com/"
                    f"{tweet_obj.user.screen_name}/status/"
                    f"{tweet_obj.id}"))
        except:
            embed.set_author(
                name='⚠️ Coronavirus Update ⚠️', url=(f"https://twitter.com/"
                f"{api.me().screen_name}"))

        # Scrub newlines from webhook URLs
        for i in range(len(webhook_urls)):
            webhook_urls[i] = webhook_urls[i].replace("\n", "")

        # Create webhook with given username and avatar, and add embed to it
        webhook = DiscordWebhook(url=webhook_urls, username=('Coronavirus '
            'Disease Tracker'), avatar_url=('https://pbs.twimg.com/'
            'profile_images/1223447558615265282/bACVLEWU_400x400.jpg'))
        webhook.add_embed(embed)

        # Send embed
        if send_flag == True:
            response = webhook.execute()

        # Logger output: Discord embed
        webhook_output = f"⚠️   Coronavirus Update   ⚠️\n\n"
        for key in stats_discord:
            webhook_output = f"{webhook_output}{key}\n{stats_discord[key]}\n\n"
        webhook_output = f"{webhook_output}{emojikey}\n\n{date_discord}"
        print(f"Discord webhook:\n{webhook_output}")

# Update Twitter bio with when last checked for case updates
def lastcheckedupdate(clock, send_flag, api, datecode):
    lastupdated = f"{clock} Updated {datecode}"

    # Define top of profile based on file
    try:
        with open ("bio_top.txt", "r") as top_file:
            bio_top=top_file.read()
            top_file.close()
    except:
        bio_top=""

    # Define bottom of profile based on file
    try:
        with open ("bio_bottom.txt", "r") as bottom_file:
            bio_bottom=bottom_file.read()
            bottom_file.close()
    except:
        bio_bottom=""

    # Create bio from top, last updated string, and bottom
    if bio_top != "":
        bio = (f"{bio_top}\n"
        f"{lastupdated}") 
    if bio_bottom != "":
        bio = (f"{bio}\n"
        f"{bio_bottom}")

    # If tweet was sent, update bio saying when last updated
    # Retry three times, if fail, raise error
    if send_flag == True:
        for attempt_no in range(0,4):
            try:
                api.update_profile(description=bio)
            except:
                if attempt_no < 3:
                    print("Retrying")
                    time.sleep(random.randint(15,30))
                else:
                    raise error

    # Logger output: Last updated timestamp
    print(bio)
